# ```setup()```

This notebook contains a method named ```setup()``` that pre-process our financial tools dataset and turn it to something usable to our models.

It have to be a very versatile method since we are going to recycle it several times for all risk envelope tools we are yet to code.

Since some of our risk envelops are predictive models (you may notice it through the libraries we import with ```import sys```), we need to make the dataset usable. 

In [1]:
import sys

!{sys.executable} -m pip install pandas numpy scikit-learn joblib


[notice] A new release of pip is available: 25.0.1 -> 26.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
import joblib
from pathlib import Path


def setup(
    csv_path,
    target_col="Price",
    horizon=1,
    train_ratio=0.7,
    val_ratio=0.15,
    save_artifacts=True,
    artifacts_path="../../data/artifacts/",
    processed_path="../../data/processed/"
):
    
    # -------------------------
    # 1. Load
    # -------------------------
    df = pd.read_csv(csv_path)
    
    # -------------------------
    # 2. Type correction
    # -------------------------
    df["Date"] = pd.to_datetime(df["Date"], format="%m/%d/%Y")
    
    if "Change %" in df.columns:
        df["Change %"] = (
            df["Change %"]
            .str.replace("%", "", regex=False)
            .astype(float) / 100
        )
    
    # -------------------------
    # 3. Sort chronologically
    # -------------------------
    df = df.sort_values("Date").reset_index(drop=True)
    
    # -------------------------
    # 4. Target engineering
    # -------------------------
    df["target_return"] = df[target_col].pct_change(horizon).shift(-horizon)
    df["target_direction"] = (df["target_return"] > 0).astype(int)
    
    # -------------------------
    # 5. Drop structural NaNs
    # -------------------------
    df = df.dropna().reset_index(drop=True)
    
    # -------------------------
    # 6. Feature / Target split
    # -------------------------
    feature_cols = [col for col in df.columns if col not in [
        "Date",
        "target_return",
        "target_direction"
    ]]
    
    X = df[feature_cols]
    y = df["target_return"]   # pode trocar para target_direction
    
    # -------------------------
    # 7. Temporal split
    # -------------------------
    n = len(df)
    train_end = int(n * train_ratio)
    val_end = int(n * (train_ratio + val_ratio))
    
    X_train, X_val, X_test = (
        X.iloc[:train_end],
        X.iloc[train_end:val_end],
        X.iloc[val_end:]
    )
    
    y_train, y_val, y_test = (
        y.iloc[:train_end],
        y.iloc[train_end:val_end],
        y.iloc[val_end:]
    )
    
    # -------------------------
    # 8. Scaling (fit only on train)
    # -------------------------
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_val_scaled = scaler.transform(X_val)
    X_test_scaled = scaler.transform(X_test)
    
    # -------------------------
    # 9. Save artifacts
    # -------------------------
    if save_artifacts:
        artifacts_dir = Path(artifacts_path)
        artifacts_dir.mkdir(parents=True, exist_ok=True)
        
        joblib.dump(scaler, artifacts_dir / "feature_scaler.pkl")
        
        processed_dir = Path(processed_path)
        processed_dir.mkdir(parents=True, exist_ok=True)
        
        df.to_csv(processed_dir / "dataset_processed.csv", index=False)
    
    return {
        "X_train": X_train_scaled,
        "X_val": X_val_scaled,
        "X_test": X_test_scaled,
        "y_train": y_train.values,
        "y_val": y_val.values,
        "y_test": y_test.values,
        "feature_columns": feature_cols
    }

Now, we upload our dataset with the financial tools into the setup method we just created:

In [3]:
data = setup(
    csv_path="../../data/processed/financial_tools_datset.csv",
    target_col="Price",
    horizon=1
)

Then we test if the cleaning, splits, and scaler base are OK.

In [4]:
print("Train shape:", data["X_train"].shape)
print("Val shape:", data["X_val"].shape)
print("Test shape:", data["X_test"].shape)

print("Número de features:", len(data["feature_columns"]))

Train shape: (956, 24)
Val shape: (205, 24)
Test shape: (206, 24)
Número de features: 24


# EPF model

Now we can finally move foward the creation of our EPF model after we have our ```setup()``` method creted

In [5]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
import numpy as np

# Target binário
y_train_bin = (data["y_train"] > 0).astype(int)
y_val_bin = (data["y_val"] > 0).astype(int)

model = LogisticRegression(max_iter=1000)
model.fit(data["X_train"], y_train_bin)

val_probs = model.predict_proba(data["X_val"])[:, 1]

auc = roc_auc_score(y_val_bin, val_probs)

print("Validation AUC:", auc)

Validation AUC: 0.9307062630877593


An AUC of 0.93 **is incredible HIGH**! It was so high, so HIGH!, the we have to test if there isn't anything wrong, like a data leaking. So we can check if there are any direct column correlation that is leaking perfect results and destroying our AUC:  

In [6]:
import numpy as np
np.corrcoef(data["X_train"].T, data["y_train"])[-1, :-1].max()

np.float64(0.05591535358276916)

The maximum correlation is 0.559, which is very low. There is no individual feature directly related to our 0.93 AUC result, so there is no occult column leaking a t+1 disguised period. Now, we test which feature has more correlation with the AUC data result:

In [7]:
pd.Series(data["feature_columns"])[
    np.argmax(
        np.abs(
            np.corrcoef(data["X_train"].T, data["y_train"])[-1, :-1]
        )
    )
]

'%K10'

It's the '%K10': our 10 period sthocastic slow oscillator. Now, lets test if there's any imbalance inflating our AUC (which can be proved for any extreme results like 0.1 or 0.9): 

In [8]:
np.mean(y_train_bin)

np.float64(0.49581589958158995)

Perfect!

- 49.6% of our returns are positive
- 50.6% are negative

This is basically a perfectly balanced target, so our AUC isn't inflated. There is not individually high correlation or inflation. 

## Training the EPF model again, but with a shuffled target

For a deeper understanding of the incredible AUC of 93% we achieved, we train the model again for a shuffled target and redo the AUC results, expecting a result as near to 0.5 as possible. 

In [9]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
import numpy as np

# Copiar para não destruir original
y_train_shuffled = y_train_bin.copy()
np.random.shuffle(y_train_shuffled)

model_shuffled = LogisticRegression(max_iter=1000)
model_shuffled.fit(data["X_train"], y_train_shuffled)

val_probs_shuffled = model_shuffled.predict_proba(data["X_val"])[:, 1]

auc_shuffled = roc_auc_score(y_val_bin, val_probs_shuffled)

print("AUC com target embaralhado:", auc_shuffled)

AUC com target embaralhado: 0.5072339615457835


After careful execution of the cell above 20 times, we got an average AUC shuffled target of 0.5155, which is great (the nearer to 0.5, the best). Now we advanced one last critical test:

# Everything here work out so well I barely can believe